types/react-query/index.d.ts

```diff
@@ -1,29 +1,46 @@
-// Type definitions for react-query 0.3
+// Type definitions for react-query 1.0
 // Project: https://github.com/tannerlinsley/react-query
 // Definitions by: Lukasz Fiszer <https://github.com/lukaszfiszer>
 //                 Jace Hensley <https://github.com/jacehensley>
 //                 Matteo Frana <https://github.com/matteofrana>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-
+// Minimum TypeScript Version: 3.0
 import { ComponentType } from 'react';
 
-// overloaded useQuery function with pagination
-export function useQuery<TResult, TVariables extends object>(
-    queryKey: QueryKey<TVariables>,
-    queryFn: QueryFunction<TResult, TVariables>,
-    options: QueryOptionsPaginated<TResult>
-): QueryResultPaginated<TResult, TVariables>;
+export type QueryKey = string | [string, ...any[]] | false | null | undefined;
+
+/**
+ * For given QueryKey get type of 2nd element of query key tuple
+ */
+export type FetchMoreVariable<T extends QueryKey> = T extends [string, infer U, ...any[]] ? U : never;
+
+/**
+ * For given QueryKey get Tuple without first element
+ */
+export type QueryKeyVariablesArgs<T extends QueryKey> = T extends any[] ? TupleTail<T> : [];
 
-export function useQuery<TResult, TVariables extends object>(
-    queryKey: QueryKey<TVariables>,
-    queryFn: QueryFunction<TResult, TVariables>,
-    options?: QueryOptions<TResult>
-): QueryResult<TResult, TVariables>;
+/**
+ * Tuple without first element
+ */
+export type TupleTail<A extends any[]> = ((...args: A) => any) extends (h: any, ...t: infer T) => any ? T : never;
 
-export type QueryKey<TVariables> = string | [string, TVariables] | false | null | QueryKeyFunction<TVariables>;
-export type QueryKeyFunction<TVariables> = () => string | [string, TVariables] | false | null;
+export function useQuery<TResult, TQueryKey extends QueryKey>(
+    queryKey: TQueryKey | (() => TQueryKey),
+    queryFn: (key: string, ...args: QueryKeyVariablesArgs<TQueryKey>) => Promise<TResult>,
+    options?: QueryOptions<TResult>,
+): QueryState<TResult>;
 
-export type QueryFunction<TResult, TVariables extends object> = (variables: TVariables) => Promise<TResult>;
+export function usePaginatedQuery<TResult, TQueryKey extends QueryKey>(
+    queryKey: TQueryKey | (() => TQueryKey),
+    queryFn: (key: string, ...args: QueryKeyVariablesArgs<TQueryKey>) => Promise<TResult>,
+    options?: QueryOptions<TResult>,
+): PaginatedQueryState<TResult>;
+
+export function useInfiniteQuery<TResult, TQueryKey extends QueryKey>(
+    queryKey: TQueryKey | (() => TQueryKey),
+    queryFn: (key: string, ...args: QueryKeyVariablesArgs<TQueryKey>) => Promise<TResult>,
+    options?: InfiniteQueryOptions<TResult, FetchMoreVariable<TQueryKey>>,
+): InfiniteQueryState<TResult, FetchMoreVariable<TQueryKey>>;
```

Comment:
Do these hooks need overloads? According to the docs the second argument can be an array of additional parameters to `queryFn` https://github.com/tannerlinsley/react-query#optional-variables

Reply:
Also can be a single object argument https://github.com/tannerlinsley/react-query#using-a-query-object-instead-of-parameters

Reply:
They definitely do need overloads, however as they might quite difficult to type correctly I planned to add them later on after this initial PR is merged